src/more_ops.rs

```diff
@@ -411,15 +440,32 @@ pub fn op_multiply(a: &mut Allocator, mut input: NodePtr, max_cost: Cost) -> Res
             continue;
         }
 
-        let (v0, l1) = int_atom(a, arg, "*")?;
-
-        total *= v0;
-        cost += MUL_COST_PER_OP;
-
-        cost += (l0 + l1) as Cost * MUL_LINEAR_COST_PER_BYTE;
-        cost += (l0 * l1) as Cost / MUL_SQUARE_COST_PER_BYTE_DIVIDER;
-
-        l0 = limbs_for_int(&total);
+        a.visit_node(arg, |node| -> Result<(), EvalErr> {
+            match node {
+                NodeVisitor::Buffer(buf) => {
+                    use crate::number::number_from_u8;
+                    total *= number_from_u8(buf);
+                    let l1 = buf.len();
+
+                    cost += MUL_COST_PER_OP;
+                    cost += (l0 + l1) as Cost * MUL_LINEAR_COST_PER_BYTE;
+                    cost += (l0 * l1) as Cost / MUL_SQUARE_COST_PER_BYTE_DIVIDER;
+                    l0 = limbs_for_int(&total);
+                    Ok(())
+                }
+                NodeVisitor::U32(val) => {
+                    total *= *val;
+                    let l1 = len_for_value(*val);
+
+                    cost += MUL_COST_PER_OP;
+                    cost += (l0 + l1) as Cost * MUL_LINEAR_COST_PER_BYTE;
+                    cost += (l0 * l1) as Cost / MUL_SQUARE_COST_PER_BYTE_DIVIDER;
+                    l0 = limbs_for_int(&total);
+                    Ok(())
```

Comment:
This is repetitive. It could be refactored by having `l1` be the output of the `match`, and in the `Pair` case, early returning. Then do the rest of this below the `match`.

Reply:
it's not so easy to "early return" in the pair case. This code is inside a callback/lambda. So early return won't return from the function. i.e. it *is* already returning.

I'll explore factoring this out though.